File: packages/client/src/sync.test.ts
import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";
import type { RequestEnvelope } from "@rayfold/server";
import { Signal, bounded } from "../../../e2e/wait.ts";
import { createBookstore } from "../../../examples/bookstore-ts/src/index.ts";
import { RayfoldCache } from "./cache.ts";
import { RayfoldClient, RayfoldClientError, type ClientOptions } from "./client.ts";
import { localStorageQueue, memoryQueue, type QueueEvent } from "./offline.ts";
import { createLocalTransport, type Transport } from "./transport.ts";

type Bookstore = ReturnType<typeof createBookstore>;
const admin = { id: "u9", role: "admin" };
const customer = { id: "u1", role: "customer" };
let bs: Bookstore;
let keyN = 0;
beforeEach(() => {
  bs = createBookstore();
  keyN = 0;
});
afterEach(() => vi.unstubAllGlobals());

/** A network that is up, down (nothing reaches the server), lossy (the server runs the batch but the answer is lost), or held at a gate. */
function network(inner: Transport) {
  const net = { mode: "up" as "up" | "down" | "lossy", sent: [] as RequestEnvelope[], gate: null as Promise<void> | null };
  const transport: Transport = {
    send(env, opts) {
      return (async function* () {
        if (net.gate) await net.gate;
        if (net.mode === "down") throw new TypeError("fetch failed");
        net.sent.push(env);
        if (net.mode === "lossy") {
          for await (const _f of inner.send(env, opts)) {
            // the server answered, and the answer never arrives
          }
          throw new TypeError("fetch failed");
        }
        yield* inner.send(env, opts);
      })();
    },
  };
  return { net, transport };
}

const clientOn = (transport: Transport, extra: Partial<ClientOptions> = {}) =>
  new RayfoldClient({ transport, keyGen: () => `sync-key-${String(++keyN).padStart(8, "0")}`, ...extra });
const stockOf = (c: RayfoldClient, id = "b1") => c.cache.get(`Book:${id}`)?.["stock"];

describe("predictions in the cache", () => {
  it("stack, and removing one leaves the server's value under whatever prediction is left", () => {
    const cache = new RayfoldCache();
    cache.applyPatch([{ set: "Book:b1", value: { stock: 5, title: "T" } }, { set: "Book:b2", value: { stock: 1 } }]);
    cache.addLayer("a", [{ set: "Book:b1", value: { stock: 6 } }]);
    cache.addLayer("b", [{ set: "Book:b1", value: { stock: 8 } }]);
    expect(cache.get("Book:b1")).toMatchObject({ stock: 8, title: "T" });
    cache.removeLayer("b");
    expect(cache.get("Book:b1")?.["stock"]).toBe(6);
    cache.applyPatch([{ set: "Book:b1", value: { stock: 7 } }]); // the server moves underneath the prediction
    expect(cache.get("Book:b1")?.["stock"]).toBe(6);
    cache.removeLayer("a");
    expect(cache.get("Book:b1")).toMatchObject({ stock: 7, title: "T" });
    expect(cache.predictions).toEqual([]);
    // guard: an entity no prediction covers takes server writes directly
    cache.applyPatch([{ set: "Book:b2", value: { stock: 2 } }]);
    expect(cache.get("Book:b2")?.["stock"]).toBe(2);
  });
});

describe("optimistic commands (sub-profile sync, spec 08 section 5)", () => {
  it("show the prediction at once, then the server's own value when it answers", async () => {
    const { net, transport } = network(createLocalTransport(bs.server, () => admin));
    const c = clientOn(transport);
    const seen = new Signal<number>();
    const stop = c.watch<{ stock: number }>("book", { id: "b1" }, { shape: "{ id stock }" }, (d) => seen.push(d.stock));
    await seen.atLeast(1, "the first value");
    let release!: () => void;
    net.gate = new Promise<void>((r) => (release = r));
    // the prediction is one short on purpose: the server's answer has to win
    const done = c.command("restock", { bookId: "b1", qty: 2 }, { shape: "{ id stock }", optimistic: (cache) => [{ set: "Book:b1", value: { stock: (cache.get("Book:b1")?.["stock"] as number) + 1 } }] });
    await seen.atLeast(2, "the prediction");
    expect(seen.items.slice(0, 2)).toEqual([5, 6]);
    expect(bs.store.books.get("b1")!.stock).toBe(5); // nothing has reached the server yet
    net.gate = null;
    release();
    await expect(done).resolves.toMatchObject({ stock: 7 });
    await seen.until((xs) => xs.at(-1) === 7, "the server's value");
    expect(stockOf(c)).toBe(7);
    expect(c.cache.predictions).toEqual([]);
    stop();
  });

  it("roll back when the command fails, and the failure still reaches the caller", async () => {
    const c = clientOn(createLocalTransport(bs.server, () => customer));
    await c.query("book", { id: "b1" }, { shape: "{ id stock }" });
    const failed = c.command("placeOrder", { input: { lines: [{ bookId: "b1", qty: 99 }] } }, { optimistic: [{ set: "Book:b1", value: { stock: -94 } }] });
    expect(stockOf(c)).toBe(-94);
    await expect(failed).rejects.toSatisfy((e: unknown) => e instanceof RayfoldClientError && e.is("OutOfStock"));
    expect(stockOf(c)).toBe(5);
    expect(c.cache.predictions).toEqual([]);
  });
});

describe("the offline queue (sub-profile sync, spec 08 section 5)", () => {
  it("keeps commands made while the server is unreachable, shows their predictions, and sends them in order with their keys", async () => {
    const { net, transport } = network(createLocalTransport(bs.server, () => admin));
    const events = new Signal<QueueEvent>();
    const c = clientOn(transport, { offline: { storage: memoryQueue() } });
    c.onQueue((e) => events.push(e));
    net.mode = "down";
    const first = c.command("restock", { bookId: "b1", qty: 1 }, { shape: "{ id stock }", optimistic: [{ set: "Book:b1", value: { stock: 6 } }] });
    await events.atLeast(1, "the first command queued");
    net.mode = "up";
    // the network is back, but a command made now still waits behind the first
    const second = c.command("restock", { bookId: "b1", qty: 2 }, { shape: "{ id stock }" });
    await events.atLeast(2, "the second command queued");
    expect(c.queued.map((q) => q.key)).toEqual(["sync-key-00000001", "sync-key-00000002"]);
    expect(stockOf(c)).toBe(6);
    expect(net.sent).toEqual([]);
    expect(bs.store.books.get("b1")!.stock).toBe(5);

    expect(await c.drain()).toBe(0);
    await expect(first).resolves.toMatchObject({ stock: 6 });
    await expect(second).resolves.toMatchObject({ stock: 8 });
    expect(net.sent.map((e) => e.ops[0]!.key)).toEqual(["sync-key-00000001", "sync-key-00000002"]);
    expect(bs.store.books.get("b1")!.stock).toBe(8);
    expect(stockOf(c)).toBe(8);
    expect(events.items.map((e) => e.type)).toEqual(["queued", "queued", "sent", "sent"]);
  });

  it("a drain while the server is still unreachable sends nothing and keeps every command", async () => {
    const { net, transport } = network(createLocalTransport(bs.server, () => admin));
    const events = new Signal<QueueEvent>();
    const c = clientOn(transport, { offline: {} });
    c.onQueue((e) => events.push(e));
    net.mode = "down";
    void c.command("restock", { bookId: "b1", qty: 1 });
    void c.command("restock", { bookId: "b2", qty: 1 });
    await events.atLeast(2, "both queued");
    // made back to back, both attempts failed; the queue holds them in the order they were made
    expect(c.queued.map((q) => q.args["bookId"])).toEqual(["b1", "b2"]);
    expect(await c.drain()).toBe(2);
    expect(bs.store.calls["Command.restock"]).toBeUndefined();
  });

  it("a command the server ran before its answer was lost is replayed on the retry, not run twice", async () => {
    const { net, transport } = network(createLocalTransport(bs.server, () => admin));
    const events = new Signal<QueueEvent>();
    const c = clientOn(transport, { offline: {} });
    c.onQueue((e) => events.push(e));
    net.mode = "lossy";
    const restocked = c.command("restock", { bookId: "b1", qty: 1 }, { shape: "{ id stock }" });
    await events.atLeast(1, "queued after the lost answer");
    expect(bs.store.calls["Command.restock"]).toBe(1); // it did run on the server
    net.mode = "up";
    expect(await c.drain()).toBe(0);
    await expect(restocked).resolves.toMatchObject({ stock: 6 });
    expect(bs.store.calls["Command.restock"]).toBe(1); // the retry carried the same key and was answered from the record
    expect(bs.store.books.get("b1")!.stock).toBe(6);
  });

  it("a queued command the server refuses is rolled back and rejected, and the next one still goes out", async () => {
    const { net, transport } = network(createLocalTransport(bs.server, () => customer));
    const events = new Signal<QueueEvent>();
    const c = clientOn(transport, { offline: {} });
    c.onQueue((e) => events.push(e));
    net.mode = "down";
    const tooMany = c.command("placeOrder", { input: { lines: [{ bookId: "b1", qty: 99 }] } }, { optimistic: [{ set: "Book:b1", value: { stock: -94 } }] });
    const one = c.command("placeOrder", { input: { lines: [{ bookId: "b1", qty: 1 }] } }, { shape: "{ id status }" });
    await events.atLeast(2, "both queued");
    net.mode = "up";
    expect(await c.drain()).toBe(0);
    await expect(tooMany).rejects.toSatisfy((e: unknown) => e instanceof RayfoldClientError && e.is("OutOfStock"));
    await expect(one).resolves.toMatchObject({ status: "PLACED" });
    expect(c.cache.predictions).toEqual([]);
    expect(events.items.map((e) => e.type)).toEqual(["queued", "queued", "failed", "sent"]);
    expect(bs.store.books.get("b1")!.stock).toBe(4);
  });

  it("the queue survives a reload: a new client with the same storage shows the prediction again and sends the command", async () => {
    const { net, transport } = network(createLocalTransport(bs.server, () => admin));
    const saved = new Map<string, string>();
    const storage = { getItem: (k: string) => saved.get(k) ?? null, setItem: (k: string, v: string) => void saved.set(k, v), removeItem: (k: string) => void saved.delete(k) };
    const events = new Signal<QueueEvent>();
    const before = clientOn(transport, { offline: { storage: localStorageQueue("orders", storage) } });
    before.onQueue((e) => events.push(e));
    net.mode = "down";
    void before.command("restock", { bookId: "b1", qty: 1 }, { shape: "{ id stock }", optimistic: [{ set: "Book:b1", value: { stock: 6 } }] });
    await events.atLeast(1, "queued");
    expect(JSON.parse(saved.get("orders")!)).toMatchObject([{ key: "sync-key-00000001", op: "restock", optimistic: [{ set: "Book:b1" }] }]);

    const after = clientOn(transport, { offline: { storage: localStorageQueue("orders", storage) } });
    expect(await after.drain()).toBe(1); // still offline, and the restored command waits
    expect(stockOf(after)).toBe(6);
    net.mode = "up";
    expect(await after.drain()).toBe(0);
    expect(net.sent.map((e) => e.ops[0]!.key)).toEqual(["sync-key-00000001"]);
    expect(bs.store.books.get("b1")!.stock).toBe(6);
    expect(saved.has("orders")).toBe(false);
  });

  it("the browser coming back online sends the waiting commands; guard: drainOnReconnect false listens for nothing", async () => {
    // Node has no window: a stand-in with a browser's addEventListener
    const window = new EventTarget();
    const listening: string[] = [];
    vi.stubGlobal("addEventListener", (type: string, fn: () => void) => {
      listening.push(type);
      window.addEventListener(type, fn);
    });
    const { net, transport } = network(createLocalTransport(bs.server, () => admin));
    const manual = clientOn(transport, { offline: { drainOnReconnect: false } });
    const c = clientOn(transport, { offline: {} });
    expect(listening).toEqual(["online"]);
    const events = new Signal<QueueEvent>();
    const manualEvents = new Signal<QueueEvent>();
    c.onQueue((e) => events.push(e));
    manual.onQueue((e) => manualEvents.push(e));
    net.mode = "down";
    const restocked = c.command("restock", { bookId: "b1", qty: 1 }, { shape: "{ id stock }" });
    const held = manual.command("restock", { bookId: "b2", qty: 1 }, { shape: "{ id stock }" });
    await events.atLeast(1, "queued while offline");
    await manualEvents.atLeast(1, "the manual client's command queued while offline");
    net.mode = "up";
    window.dispatchEvent(new Event("online"));
    expect(await bounded(restocked, "sent on the online event")).toEqual({ $type: "Book", id: "b1", stock: 6 });
    expect(events.items.map((e) => e.type)).toEqual(["queued", "sent"]);
    expect(manual.queued.map((q) => q.args["bookId"])).toEqual(["b2"]);
    // the manual client sends only when told to
    expect(await manual.drain()).toBe(0);
    expect(await held).toEqual({ $type: "Book", id: "b2", stock: 3 });
    expect(net.sent.map((e) => e.ops[0]!.args?.["bookId"])).toEqual(["b1", "b2"]);
  });

  it("guard: without `offline`, a failed network rejects as before, rolls the prediction back, and queues nothing", async () => {
    const { net, transport } = network(createLocalTransport(bs.server, () => admin));
    const c = clientOn(transport);
    net.mode = "down";
    await expect(c.command("restock", { bookId: "b1", qty: 1 }, { optimistic: [{ set: "Book:b1", value: { stock: 6 } }] })).rejects.toThrow("fetch failed");
    expect(c.queued).toEqual([]);
    expect(c.cache.get("Book:b1")).toBeUndefined();
    expect(await c.drain()).toBe(0);
  });
});

File: kotlin/rayfold-client/src/test/kotlin/dev/rayfold/client/SyncTest.kt
package dev.rayfold.client

import dev.rayfold.core.RayfoldException
import dev.rayfold.core.RayfoldServer
import dev.rayfold.core.Resolvers
import dev.rayfold.core.SchemaText
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.async
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.emitAll
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withTimeout
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.int
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir
import java.io.File
import java.net.ConnectException
import java.net.SocketException
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.atomic.AtomicInteger
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * Sub-profile `sync` in the Kotlin client, against the real Kotlin server through a network that can be up, down,
 * lossy (the server runs the batch and the answer is lost) or held at a gate. Every test is bounded to 5 s.
 */
class SyncTest {
    private val schema = """
        entity Book { id: ID title: String stock: Int }
        error OutOfStock { available: Int }
        query book(id: ID): Book?
        command restock(id: ID, qty: Int): Book
        command buy(id: ID, qty: Int): Book throws OutOfStock
    """
    private val books = ConcurrentHashMap(mapOf("b1" to book("b1", 3), "b2" to book("b2", 1)))
    private val restocks = AtomicInteger()
    private val viewer = buildJsonObject { put("id", "u1") }
    private val server = RayfoldServer(
        SchemaText.load(schema).ir,
        Resolvers(
            queries = mapOf("book" to { a, _ -> books[a.id()] }),
            commands = mapOf(
                "restock" to { a, _ ->
                    restocks.incrementAndGet()
                    book(a.id(), stockOf(a.id()) + a.qty()).also { books[a.id()] = it }
                },
                "buy" to { a, _ ->
                    val stock = stockOf(a.id())
                    if (a.qty() > stock) throw RayfoldException.domain("OutOfStock", buildJsonObject { put("available", stock) }, "Only $stock left")
                    book(a.id(), stock - a.qty()).also { books[a.id()] = it }
                },
            ),
        ),
    )

    private fun book(id: String, stock: Int) = buildJsonObject { put("id", id); put("title", "T$id"); put("stock", stock) }
    private fun stockOf(id: String) = books[id]?.get("stock")?.jsonPrimitive?.int ?: 0
    private fun JsonObject.id() = this["id"]?.jsonPrimitive?.content ?: error("no id")
    private fun JsonObject.qty() = this["qty"]?.jsonPrimitive?.int ?: error("no qty")
    private fun JsonElement.stock() = jsonObject["stock"]?.jsonPrimitive?.int

    private inner class Network : Transport {
        @Volatile var mode = "up"
        @Volatile var gate: CompletableDeferred<Unit>? = null
        val sent = CopyOnWriteArrayList<JsonObject>()

        override fun send(envelope: JsonObject, safe: Boolean): Flow<JsonObject> = flow {
            gate?.await()
            if (mode == "down") throw ConnectException("Connection refused")
            sent.add(envelope)
            if (mode == "lossy") {
                server.execute(envelope, viewer).collect { } // the server answered, and the answer never arrives
                throw SocketException("Connection reset")
            }
            emitAll(server.execute(envelope, viewer))
        }

        fun keys() = sent.map { it["ops"]?.let { ops -> (ops as kotlinx.serialization.json.JsonArray)[0].jsonObject["key"]?.jsonPrimitive?.content } }
    }

    private val keyN = AtomicInteger()
    private fun client(net: Network, offline: OfflineOptions? = null) =
        RayfoldClient(net, ClientOptions(keyGen = { "sync-key-" + keyN.incrementAndGet().toString().padStart(8, '0') }, offline = offline))

    private fun bounded(block: suspend CoroutineScope.() -> Unit) = runBlocking { withTimeout(5_000) { block() } }
    private fun predict(id: String, stock: Int) = listOf(OptimisticOp("Book:$id", buildJsonObject { put("stock", stock) }))
    private fun RayfoldClient.stock(id: String = "b1") = cache.get("Book:$id")?.get("stock")?.jsonPrimitive?.int

    /** Queue events as they happen, to wait on. */
    private fun events(c: RayfoldClient): Channel<QueueEvent> = Channel<QueueEvent>(Channel.UNLIMITED).also { ch -> c.onQueue { ch.trySend(it) } }

    @Test
    fun `predictions stack, and removing one leaves the server's value under whatever prediction is left`() {
        val cache = RayfoldCache()
        cache.applyPatch(listOf(buildJsonObject { put("set", "Book:b1"); put("value", buildJsonObject { put("stock", 5); put("title", "T") }) }))
        cache.addLayer("a", predict("b1", 6))
        cache.addLayer("b", predict("b1", 8))
        assertEquals(8, cache.get("Book:b1")?.get("stock")?.jsonPrimitive?.int)
        cache.removeLayer("b")
        assertEquals(6, cache.get("Book:b1")?.get("stock")?.jsonPrimitive?.int)
        cache.applyPatch(listOf(buildJsonObject { put("set", "Book:b1"); put("value", buildJsonObject { put("stock", 7) }) })) // the server moves underneath
        assertEquals(6, cache.get("Book:b1")?.get("stock")?.jsonPrimitive?.int)
        cache.removeLayer("a")
        assertEquals(7, cache.get("Book:b1")?.get("stock")?.jsonPrimitive?.int)
        assertEquals(JsonPrimitive("T"), cache.get("Book:b1")?.get("title"))
        assertEquals(emptyList(), cache.predictions)
    }

    @Test
    fun `a command shows its prediction at once, then the server's own value when it answers`() = bounded {
        val net = Network()
        val c = client(net)
        val seen = Channel<Int>(Channel.UNLIMITED)
        val watching = launch { c.watch("book", args("id" to "b1"), "{ id stock }").collect { seen.send(it.stock() ?: -1) } }
        assertEquals(3, seen.receive())
        net.gate = CompletableDeferred()
        // the prediction is one short on purpose: the server's answer has to win
        val done = async { c.command("restock", args("id" to "b1", "qty" to 2), "{ id stock }", optimistic = predict("b1", 4)) }
        assertEquals(4, seen.receive())
        assertEquals(3, stockOf("b1"), "nothing has reached the server yet")
        net.gate?.complete(Unit)
        assertEquals(5, done.await().stock())
        // the answer lands under the prediction, which the watch reports as a change to its book (still 4), and then the
        // settled command takes the prediction away; nothing else follows
        assertEquals(listOf(4, 5), listOf(seen.receive(), seen.receive()))
        assertTrue(seen.tryReceive().isFailure, "no value after the server's own")
        assertEquals(5, c.stock())
        assertEquals(emptyList(), c.cache.predictions)
        watching.cancel()
    }

    @Test
    fun `a failed command rolls its prediction back, and the failure still reaches the caller`() = bounded {
        val c = client(Network())
        c.query("book", args("id" to "b1"), "{ id stock }")
        val e = assertFailsWith<RayfoldClientException> { c.command("buy", args("id" to "b1", "qty" to 99), optimistic = predict("b1", -96)) }
        assertTrue(e.isType("OutOfStock"))
        assertEquals(3, c.stock())
        assertEquals(emptyList(), c.cache.predictions)
    }

    @Test
    fun `commands made while the server is unreachable wait with their predictions and go out in order with their keys`() = bounded {
        val net = Network()
        val c = client(net, OfflineOptions())
        val events = events(c)
        net.mode = "down"
        val first = async { c.command("restock", args("id" to "b1", "qty" to 1), "{ id stock }", optimistic = predict("b1", 4)) }
        assertEquals(QueueEvent.Type.QUEUED, events.receive().type)
        // a second command while the server is still out waits behind the first
        val second = async { c.command("restock", args("id" to "b1", "qty" to 2), "{ id stock }") }
        assertEquals(QueueEvent.Type.QUEUED, events.receive().type)
        assertEquals(listOf("sync-key-00000001", "sync-key-00000002"), c.queued.map { it.key })
        assertEquals(4, c.stock())
        assertEquals(emptyList(), net.sent)

        net.mode = "up"
        assertEquals(0, c.drain())
        assertEquals(4, first.await().stock())
        assertEquals(6, second.await().stock())
        assertEquals(listOf("sync-key-00000001", "sync-key-00000002"), net.keys())
        assertEquals(6, stockOf("b1"))
        assertEquals(6, c.stock())
    }

    @Test
    fun `a drain while the server is still unreachable sends nothing and keeps every command`() = bounded {
        val net = Network()
        val c = client(net, OfflineOptions())
        val events = events(c)
        net.mode = "down"
        launch { c.command("restock", args("id" to "b1", "qty" to 1)) }
        events.receive()
        launch { c.command("restock", args("id" to "b2", "qty" to 1)) }
        events.receive()
        assertEquals(2, c.drain())
        assertEquals(0, restocks.get())
        coroutineContext.cancelChildren()
    }

    @Test
    fun `a command the server ran before its answer was lost is replayed on the retry, not run twice`() = bounded {
        val net = Network()
        val c = client(net, OfflineOptions())
        val events = events(c)
        net.mode = "lossy"
        val restocked = async { c.command("restock", args("id" to "b1", "qty" to 1), "{ id stock }") }
        events.receive()
        assertEquals(1, restocks.get(), "it did run on the server")
        net.mode = "up"
        assertEquals(0, c.drain())
        assertEquals(4, restocked.await().stock())
        assertEquals(1, restocks.get(), "the retry carried the same key and was answered from the record")
    }

    @Test
    fun `a queued command the server refuses is rolled back and refused, and the next one still goes out`() = bounded {
        val net = Network()
        val c = client(net, OfflineOptions())
        val events = events(c)
        net.mode = "down"
        val tooMany = async { runCatching { c.command("buy", args("id" to "b1", "qty" to 99), optimistic = predict("b1", -96)) } }
        events.receive()
        val one = async { c.command("buy", args("id" to "b1", "qty" to 1), "{ id stock }") }
        events.receive()
        net.mode = "up"
        assertEquals(0, c.drain())
        assertTrue((tooMany.await().exceptionOrNull() as? RayfoldClientException)?.isType("OutOfStock") == true)
        assertEquals(2, one.await().stock())
        assertEquals(listOf(QueueEvent.Type.FAILED, QueueEvent.Type.SENT), listOf(events.receive().type, events.receive().type))
        assertEquals(emptyList(), c.cache.predictions)
    }

    @Test
    fun `the queue survives a restart in a file, shows its prediction again, and sends the command`(@TempDir dir: File) = bounded {
        val net = Network()
        val file = File(dir, "queue.json")
        val before = client(net, OfflineOptions(FileQueueStorage(file)))
        val events = events(before)
        net.mode = "down"
        launch { before.command("restock", args("id" to "b1", "qty" to 1), "{ id stock }", optimistic = predict("b1", 4)) }
        events.receive()
        assertTrue(file.readText().contains("sync-key-00000001"))

        val after = client(net, OfflineOptions(FileQueueStorage(file)))
        assertEquals(1, after.drain(), "still offline: the restored command waits")
        assertEquals(4, after.stock())
        net.mode = "up"
        assertEquals(0, after.drain())
        assertEquals(listOf("sync-key-00000001"), net.keys())
        assertEquals(4, stockOf("b1"))
        assertTrue(!file.exists(), "an empty queue leaves no file")
        coroutineContext.cancelChildren()
    }

    @Test
    fun `a command made once the server is back sends the waiting ones first and then itself, with no drain`() = bounded {
        val net = Network()
        val c = client(net, OfflineOptions())
        val events = events(c)
        net.mode = "down"
        val first = async { c.command("restock", args("id" to "b1", "qty" to 1), "{ id stock }") }
        assertEquals(QueueEvent.Type.QUEUED, events.receive().type)
        net.mode = "up"
        assertEquals(6, c.command("restock", args("id" to "b1", "qty" to 2), "{ id stock }").stock())
        assertEquals(4, first.await().stock())
        assertEquals(listOf("sync-key-00000001", "sync-key-00000002"), net.keys())
        assertEquals(emptyList(), c.queued)
    }

    @Test
    fun `after a restart the first command sends the restored ones before itself`(@TempDir dir: File) = bounded {
        val net = Network()
        val file = File(dir, "queue.json")
        val before = client(net, OfflineOptions(FileQueueStorage(file)))
        val events = events(before)
        net.mode = "down"
        launch { before.command("restock", args("id" to "b1", "qty" to 1), "{ id stock }", optimistic = predict("b1", 4)) }
        events.receive()
        coroutineContext.cancelChildren()

        net.mode = "up"
        val after = client(net, OfflineOptions(FileQueueStorage(file)))
        assertEquals(4, after.stock(), "the restored prediction is shown")
        assertEquals(emptyList(), net.sent)
        assertEquals(6, after.command("restock", args("id" to "b1", "qty" to 2), "{ id stock }").stock())
        assertEquals(listOf("sync-key-00000001", "sync-key-00000002"), net.keys())
        assertEquals(emptyList(), after.queued)
        assertTrue(!file.exists())
    }

    @Test
    fun `a damaged queue file is dropped instead of blocking the client`(@TempDir dir: File) {
        val file = File(dir, "queue.json").apply { writeText("{ not json") }
        assertEquals(emptyList(), FileQueueStorage(file).load())
    }

    @Test
    fun `guard - without offline, an unreachable server fails the command, rolls the prediction back and queues nothing`() = bounded {
        val net = Network()
        val c = client(net)
        net.mode = "down"
        assertFailsWith<ConnectException> { c.command("restock", args("id" to "b1", "qty" to 1), optimistic = predict("b1", 4)) }
        assertEquals(emptyList(), c.queued)
        assertNull(c.cache.get("Book:b1"))
        assertEquals(0, c.drain())
    }

    private fun kotlin.coroutines.CoroutineContext.cancelChildren() = this[kotlinx.coroutines.Job]?.children?.forEach { it.cancel() }
}
